lib/acts_as_tenant/model_extensions.rb

```diff
@@ -20,39 +20,43 @@ def acts_as_tenant(tenant = :account, **options)
           belongs_to tenant, **valid_options
         end
 
-        default_scope lambda {
-          if ActsAsTenant.should_require_tenant? && ActsAsTenant.current_tenant.nil? && !ActsAsTenant.unscoped?
-            raise ActsAsTenant::Errors::NoTenantSet
-          end
+        unless ActsAsTenant.already_scoped? self
+          default_scope lambda {
+            if ActsAsTenant.should_require_tenant? && ActsAsTenant.current_tenant.nil? && !ActsAsTenant.unscoped?
+              raise ActsAsTenant::Errors::NoTenantSet
+            end
 
-          if ActsAsTenant.current_tenant
-            keys = [ActsAsTenant.current_tenant.send(pkey)].compact
-            keys.push(nil) if ActsAsTenant.has_global_records?(self.klass)
-
-            if options[:through]
-              query_criteria = {options[:through] => {fkey.to_sym => keys}}
-              query_criteria[polymorphic_type.to_sym] = ActsAsTenant.current_tenant.class.to_s if options[:polymorphic]
-              joins(options[:through])
-                .where(query_criteria)
-                .ids
-                .then do |ids|
-                  if options[:unscoped]
-                    where(id: ids).or(where(options[:unscoped]))
-                  else
-                    where(id: ids)
+            if ActsAsTenant.current_tenant
+              keys = [ActsAsTenant.current_tenant.send(pkey)].compact
+              keys.push(nil) if ActsAsTenant.has_global_records?(klass)
+
+              if options[:through]
+                query_criteria = {options[:through] => {fkey.to_sym => keys}}
+                query_criteria[polymorphic_type.to_sym] = ActsAsTenant.current_tenant.class.to_s if options[:polymorphic]
+                joins(options[:through])
+                  .where(query_criteria)
+                  .ids
+                  .then do |ids|
+                    if options[:unscoped]
+                      where(id: ids).or(where(options[:unscoped]))
+                    else
+                      where(id: ids)
+                    end
                   end
+              else
+                query_criteria = {fkey.to_sym => keys}
+                query_criteria[polymorphic_type.to_sym] = ActsAsTenant.current_tenant.class.to_s if options[:polymorphic]
+                where(query_criteria).then do |q|
+                  options[:unscoped] ? q.or(where(options[:unscoped])) : q
                 end
-            else
-              query_criteria = {fkey.to_sym => keys}
-              query_criteria[polymorphic_type.to_sym] = ActsAsTenant.current_tenant.class.to_s if options[:polymorphic]
-              where(query_criteria).then do |q|
-                options[:unscoped] ? q.or(where(options[:unscoped])) : q
               end
+            else
+              all
             end
-          else
-            all
-          end
-        }
+          }
+        end
+
+        ActsAsTenant.add_already_scoped self
```

Comment:
what's the advantage of this being outside of `unless ActsAsTenant.already_scoped?` feels like it'd just add more models to the directory than necessary, increasing the memory footprint a bit.

Reply:
The 'directory' won't include duplicates if that's what you mean? The directory is a hash rather than array so as to avoid an include/ find on every check.

The idea is that every model we've called `acts_as_tenant` on is stored in this directory, otherwise how we know that the `default_scope` was already applied?

Reply:
Fair, no particular reason that I can remember/ see. Have changed now.